# STOR 120 - Lab 2: Numbers, Table Operations, & Arrays

Welcome to Lab 2!  In this lab you will:
* work with names, numbers, and expressions
* import a module and practice table operations
* create tables and practice analyzing them with your knowledge of table operations and arrays

# 1. Numbers

Quantitative information arises everywhere in data science. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number 3.25. (Run the cell and see.)

In [126]:
3.2500

3.25

Notice that we didn't have to `print`. When you run a notebook cell, if the last line has a value, then Jupyter helpfully prints out that value for you. However, it won't print out prior lines automatically.

In [127]:
print(2)
3
4

2


4

Above, you should see that 4 is the value of the last expression, 2 is printed, but 3 is lost forever because it was neither printed nor last.

You don't want to print everything all the time anyway.  But if you feel sorry for 3, change the cell above to print it.

## 1.1. Arithmetic
The line in the next cell subtracts.  Its value is what you'd expect.  Run it.

In [128]:
3.25 - 1.5

1.75

Many basic arithmetic operations are built into Python.  The textbook section on [Expressions](https://inferentialthinking.com/chapters/03/1/Expressions.html) describes all the arithmetic operators used in the course.  The common operator that differs from typical math notation is `**`, which raises one number to the power of the other. So, `2**3` stands for $2^3$ and evaluates to 8. 

The order of operations is the same as what you learned in elementary school, and Python also has parentheses.  For example, compare the outputs of the cells below.

In [129]:
6+6*5-6*3**2*2**3/4*7

-720.0

In [130]:
6+(6*5-(6*3))**2*((2**3)/4*7)

2022.0

In standard math notation, the first expression is

$$6 + 6 \times 5 - 6 \times 3^2 \times \frac{2^3}{4} \times 7,$$

while the second expression is

$$6 + (6 \times 5 - (6 \times 3))^2 \times (\frac{(2^3)}{4} \times 7).$$

**Question 1.1.1.** Write a Python expression in this next cell that's equal to $11 \times  2^{0.5 } - \frac{7}{33} + 3$.

Replace the ellipses (`...`) with your expression.  Try to use parentheses only when necessary.

In [131]:
11 * (2)**(1/2) - (7/33) + 3

18.344227973982832

# 2. Names
In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [132]:
ten = 3 * 2 + 4

When you run that cell, Python first computes the value of the expression on the right-hand side, `3 * 2 + 4`, which is the number 10.  Then it assigns that value to the name `ten`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `ten`:

In [133]:
ten

10

The statement `ten = 3 * 2 + 4` is not asserting that `ten` is already equal to `3 * 2 + 4`, as we might expect by analogy with math notation.  Rather, that line of code changes what `ten` means; it now refers to the value 10, whereas before it meant nothing at all.

If the designers of Python had been ruthlessly pedantic, they might have made us write

    define the name ten to hereafter have the value of 3 * 2 + 4 

instead.  You will probably appreciate the brevity of "`=`"!  But keep in mind that this is the real meaning.

**Question 2.1.** Try writing code that uses a name (like `eleven`) that hasn't been assigned to anything.  You'll see an error! Leave this error in your submitted notebook.

In [134]:
print(eleven)

NameError: name 'eleven' is not defined

A common pattern in Jupyter notebooks is to assign a value to a name and then immediately evaluate the name in the last line in the cell so that the value is displayed as output. 

In [135]:
close_to_pi = 355/113
close_to_pi

3.1415929203539825

## 2.1. Computations with Names

Another common pattern is that a series of lines in a single cell will build up a complex computation in stages, naming the intermediate results.

In [136]:
semimonthly_salary = 2205.78
monthly_salary = 2 * semimonthly_salary
number_of_months_in_a_year = 12
yearly_salary = number_of_months_in_a_year * monthly_salary
yearly_salary

52938.72

Names in Python can have letters (upper- and lower-case letters are both okay and count as different letters), underscores, and numbers.  The first character can't be a number (otherwise a name might look like a number).  And names can't contain spaces, since spaces are used to separate pieces of code from each other.

Other than those rules, what you name something doesn't matter *to Python*.  For example, this cell does the same thing as the above cell, except everything has a different name:

In [137]:
a = 2205.78
b = 2 * a
c = 12
d = c * b
d

52938.72

**However**, names are very important for making your code *readable* to yourself and others.  The cell above is shorter, but it's totally useless without an explanation of what it does.

**Question 2.1.1.** Assign the name `minutes_in_a_decade` to the number of minutes between midnight January 1, 2016 and midnight January 1, 2026. Note that there are leap years in this span of a decade. A non-leap year has 365 days and a leap year has 366 days.

In [138]:
# Change the next line
# so that it computes the number of minutes in a decade
# and assigns that number the name minutes_in_a_decade.

regular_years = 8
leap_years = 2

total_days = (regular_years * 365) + (leap_years * 366)
hours_per_day = 24
minutes_per_hour = 60

minutes_in_a_decade = total_days * hours_per_day * minutes_per_hour

# We've put this line in this cell
# so that it will print the value you've given to minutes_in_a_decade when you run it.
# You don't need to change this.
minutes_in_a_decade

5258880

# 2.2. Comments
You may have noticed these lines in the cell in which you answered Question 2.1.1:

    # Change the next line 
    # so that it computes the number of minutes in a decade 
    # and assigns that number the name, minutes_in_a_decade.
    
This is called a *comment*. It doesn't make anything happen in Python; Python ignores anything on a line after a `#`.  Instead, it's there to communicate something about the code to you, the human reader. Comments are extremely useful. 


<img src="http://imgs.xkcd.com/comics/future_self.png">
Source: http://imgs.xkcd.com/comics/future_self.png

# 3. Table operations

The table `farmers_markets.csv` contains data on farmers' markets in the United States  (data collected by the USDA).  Each row represents one such market.

Run the next cell to load the `farmers_markets` table.

In [139]:
# Just run this cell
# The farmers market csv file must be saved in the same directory as this lab!

from datascience import *
farmers_markets = Table.read_table('../../shared/Data/farmers_markets.csv')
farmers_markets

FMID,MarketName,street,city,County,State,zip,x,y,Website,Facebook,Twitter,Youtube,OtherMedia,Organic,Tofu,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,WildHarvested,updateTime,Location,Credit,WIC,WICcash,SFMNP,SNAP,Season1Date,Season1Time,Season2Date,Season2Time,Season3Date,Season3Time,Season4Date,Season4Time
1012063,Caledonia Farmers Market Association - Danville,nan,Danville,Caledonia,Vermont,05828,-72.1403,44.411,https://sites.google.com/site/caledoniafarmersmarket/,https://www.facebook.com/Danville.VT.Farmers.Market/,nan,nan,nan,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,Y,Y,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,6/28/2016 12:10:09 PM,nan,Y,Y,N,Y,N,06/08/2016 to 10/12/2016,Wed: 9:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011871,Stearns Homestead Farmers' Market,6975 Ridge Road,Parma,Cuyahoga,Ohio,44130,-81.7286,41.3751,http://Stearnshomestead.com,nan,nan,nan,nan,-,N,Y,N,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,N,N,N,N,N,N,N,Y,N,N,N,Y,N,4/9/2016 8:05:17 PM,nan,Y,Y,N,Y,Y,06/25/2016 to 10/01/2016,Sat: 9:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011878,100 Mile Market,507 Harrison St,Kalamazoo,Kalamazoo,Michigan,49007,-85.5749,42.296,http://www.pfcmarkets.com,https://www.facebook.com/100MileMarket/?fref=ts,nan,nan,https://www.instagram.com/100milemarket/,N,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,Y,N,N,Y,Y,N,N,N,N,4/16/2016 12:37:56 PM,nan,Y,Y,N,Y,Y,05/04/2016 to 10/12/2016,Wed: 3:00 PM-7:00 PM;,nan,nan,nan,nan,nan,nan
1009364,106 S. Main Street Farmers Market,106 S. Main Street,Six Mile,nan,South Carolina,29682,-82.8187,34.8042,http://thetownofsixmile.wordpress.com/,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2013,nan,Y,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
1010691,10th Steet Community Farmers Market,10th Street and Poplar,Lamar,Barton,Missouri,64759,-94.2746,37.4956,nan,nan,nan,nan,http://agrimissouri.com/mo-grown/grodetail.php?type=mo-g ...,-,N,Y,N,Y,N,Y,N,Y,Y,Y,Y,N,Y,N,N,Y,Y,Y,Y,N,N,N,N,Y,N,N,N,N,N,10/28/2014 9:49:46 AM,nan,Y,N,N,N,N,04/02/2014 to 11/30/2014,Wed: 3:00 PM-6:00 PM;Sat: 8:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1002454,112st Madison Avenue,112th Madison Avenue,New York,New York,New York,10029,-73.9493,40.7939,nan,nan,nan,nan,nan,-,N,Y,N,Y,Y,N,N,Y,Y,Y,Y,N,N,N,Y,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,3/1/2012 10:38:22 AM,Private business parking lot,N,N,Y,Y,N,July to November,Tue:8:00 am - 5:00 pm;Sat:8:00 am - 8:00 pm;,nan,nan,nan,nan,nan,nan
1011100,12 South Farmers Market,3000 Granny White Pike,Nashville,Davidson,Tennessee,37204,-86.7907,36.1184,http://www.12southfarmersmarket.com,12_South_Farmers_Market,@12southfrmsmkt,nan,@12southfrmsmkt,Y,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,N,Y,N,Y,N,Y,Y,Y,N,5/1/2015 10:40:56 AM,nan,Y,N,N,N,Y,05/05/2015 to 10/27/2015,Tue: 3:30 PM-6:30 PM;,nan,nan,nan,nan,nan,nan
1009845,125th Street Fresh Connect Farmers' Market,"163 West 125th Street and Adam Clayton Powell, Jr. Blvd.",New York,New York,New York,10027,-73.9482,40.809,http://www.125thStreetFarmersMarket.com,https://www.facebook.com/125thStreetFarmersMarket,https://twitter.com/FarmMarket125th,nan,Instagram--> 125thStreetFarmersMarket,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,Y,Y,N,Y,N,Y,N,N,N,4/7/2014 4:32:01 PM,Federal/State government building grounds,Y,Y,N,Y,Y,06/10/2014 to 11/25/2014,Tue: 10:00 AM-7:00 PM;,nan,nan,nan,nan,nan,nan
1005586,12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,nan,https://www.facebook.com/pages/12th-Brandywine-Urban-Far ...,nan,nan,https://www.facebook.com/delawareurbanfarmcoalition,N,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,4/3/2014 3:43:31 PM,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,05/16/2014 to 10/17/2014,Fri: 8:00 AM-11:00 AM;,nan,nan,nan,nan,nan,nan
1008071,14&U Farmers' Market,1400 U Street NW,Washington,District of Columbia,District of Columbia,20009,

### `select`

Most of the columns are about particular products -- whether the market sells tofu, pet food, etc.  If we're not interested in that information, it just makes the table difficult to read.  This comes up more than you might think, because people who collect and publish data may not know ahead of time what people will want to do with it.

In such situations, we can use the table method `select` to choose only the columns that we want in a particular table. It takes any number of arguments. Each should be the name of a column in the table. It returns a new table with only those columns in it. The columns are in the order *in which they were listed as arguments*.

For example, the value of `farmers_markets.select("MarketName", "State")` is a table with only the name and the state of each farmers' market in `farmers_markets`.



**Question 3.1.** Use `select` to create a table with only the markets' names, cities, counties, states, zip codes (`zip`), latitude (`x`), and longitude (`y`).  Call that new table `farmers_markets_locations`.

*Hint:* Make sure to be exact when using column names with `select`; double-check capitalization!

In [140]:
farmers_markets_locations = farmers_markets.select("MarketName", "city", "County", "State", "zip", "x", "y")
farmers_markets_locations

MarketName,city,County,State,zip,x,y
Caledonia Farmers Market Association - Danville,Danville,Caledonia,Vermont,05828,-72.1403,44.411
Stearns Homestead Farmers' Market,Parma,Cuyahoga,Ohio,44130,-81.7286,41.3751
100 Mile Market,Kalamazoo,Kalamazoo,Michigan,49007,-85.5749,42.296
106 S. Main Street Farmers Market,Six Mile,nan,South Carolina,29682,-82.8187,34.8042
10th Steet Community Farmers Market,Lamar,Barton,Missouri,64759,-94.2746,37.4956
112st Madison Avenue,New York,New York,New York,10029,-73.9493,40.7939
12 South Farmers Market,Nashville,Davidson,Tennessee,37204,-86.7907,36.1184
125th Street Fresh Connect Farmers' Market,New York,New York,New York,10027,-73.9482,40.809
12th & Brandywine Urban Farm Market,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421
14&U Farmers' Market,Washington,District of Columbia,District of Columbia,20009,-77.0321,38.917


### `drop`

`drop` serves the same purpose as `select`, but it takes away the columns that you provide rather than the ones that you don't provide. Like `select`, `drop` returns a new table.

**Question 3.2.** Suppose you just didn't want the `FMID` and `updateTime` columns in `farmers_markets`.  Create a table that's a copy of `farmers_markets` but doesn't include those columns.  Call that table `farmers_markets_without_fmid`.

In [141]:
farmers_markets_without_fmid = farmers_markets.drop("FMID", "updateTime")
farmers_markets_without_fmid

MarketName,street,city,County,State,zip,x,y,Website,Facebook,Twitter,Youtube,OtherMedia,Organic,Tofu,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,WildHarvested,Location,Credit,WIC,WICcash,SFMNP,SNAP,Season1Date,Season1Time,Season2Date,Season2Time,Season3Date,Season3Time,Season4Date,Season4Time
Caledonia Farmers Market Association - Danville,nan,Danville,Caledonia,Vermont,05828,-72.1403,44.411,https://sites.google.com/site/caledoniafarmersmarket/,https://www.facebook.com/Danville.VT.Farmers.Market/,nan,nan,nan,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,Y,Y,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,nan,Y,Y,N,Y,N,06/08/2016 to 10/12/2016,Wed: 9:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
Stearns Homestead Farmers' Market,6975 Ridge Road,Parma,Cuyahoga,Ohio,44130,-81.7286,41.3751,http://Stearnshomestead.com,nan,nan,nan,nan,-,N,Y,N,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,N,N,N,N,N,N,N,Y,N,N,N,Y,N,nan,Y,Y,N,Y,Y,06/25/2016 to 10/01/2016,Sat: 9:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
100 Mile Market,507 Harrison St,Kalamazoo,Kalamazoo,Michigan,49007,-85.5749,42.296,http://www.pfcmarkets.com,https://www.facebook.com/100MileMarket/?fref=ts,nan,nan,https://www.instagram.com/100milemarket/,N,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,Y,N,N,Y,Y,N,N,N,N,nan,Y,Y,N,Y,Y,05/04/2016 to 10/12/2016,Wed: 3:00 PM-7:00 PM;,nan,nan,nan,nan,nan,nan
106 S. Main Street Farmers Market,106 S. Main Street,Six Mile,nan,South Carolina,29682,-82.8187,34.8042,http://thetownofsixmile.wordpress.com/,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,nan,Y,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
10th Steet Community Farmers Market,10th Street and Poplar,Lamar,Barton,Missouri,64759,-94.2746,37.4956,nan,nan,nan,nan,http://agrimissouri.com/mo-grown/grodetail.php?type=mo-g ...,-,N,Y,N,Y,N,Y,N,Y,Y,Y,Y,N,Y,N,N,Y,Y,Y,Y,N,N,N,N,Y,N,N,N,N,N,nan,Y,N,N,N,N,04/02/2014 to 11/30/2014,Wed: 3:00 PM-6:00 PM;Sat: 8:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
112st Madison Avenue,112th Madison Avenue,New York,New York,New York,10029,-73.9493,40.7939,nan,nan,nan,nan,nan,-,N,Y,N,Y,Y,N,N,Y,Y,Y,Y,N,N,N,Y,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,Private business parking lot,N,N,Y,Y,N,July to November,Tue:8:00 am - 5:00 pm;Sat:8:00 am - 8:00 pm;,nan,nan,nan,nan,nan,nan
12 South Farmers Market,3000 Granny White Pike,Nashville,Davidson,Tennessee,37204,-86.7907,36.1184,http://www.12southfarmersmarket.com,12_South_Farmers_Market,@12southfrmsmkt,nan,@12southfrmsmkt,Y,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,N,Y,N,Y,N,Y,Y,Y,N,nan,Y,N,N,N,Y,05/05/2015 to 10/27/2015,Tue: 3:30 PM-6:30 PM;,nan,nan,nan,nan,nan,nan
125th Street Fresh Connect Farmers' Market,"163 West 125th Street and Adam Clayton Powell, Jr. Blvd.",New York,New York,New York,10027,-73.9482,40.809,http://www.125thStreetFarmersMarket.com,https://www.facebook.com/125thStreetFarmersMarket,https://twitter.com/FarmMarket125th,nan,Instagram--> 125thStreetFarmersMarket,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,Y,Y,N,Y,N,Y,N,N,N,Federal/State government building grounds,Y,Y,N,Y,Y,06/10/2014 to 11/25/2014,Tue: 10:00 AM-7:00 PM;,nan,nan,nan,nan,nan,nan
12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,nan,https://www.facebook.com/pages/12th-Brandywine-Urban-Far ...,nan,nan,https://www.facebook.com/delawareurbanfarmcoalition,N,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,05/16/2014 to 10/17/2014,Fri: 8:00 AM-11:00 AM;,nan,nan,nan,nan,nan,nan
14&U Farmers' Market,1400 U Street NW,Washington,District of Columbia,District of Columbia,20009,-77.0321,38.917,nan,https://www.facebook.com/14UFarmersMarket,https://twitter.com/14UFarmersMkt,nan,nan,Y,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,N,N,N,N,Y,Y,Y,Y,N,N,N,Other,Y,Y,Y,Y,Y,05/03/2014 to 11/22/2014,Sat: 9:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan


Now, suppose we want to answer some questions about farmers' markets in the US. For example, which market(s) are the farthest north (given by the largest values in the `y` column)? 

To answer this, we'll sort `farmers_markets_locations` by latitude.

In [142]:
farmers_markets_locations.sort('y')

MarketName,city,County,State,zip,x,y
Anne Heyliger Vegetable Market,Saint Croix,Frederiksted,Virgin Islands,nan,-64.8799,17.7099
La Reine Farmers Market,Saint Croix,Christiansted,Virgin Islands,nan,-64.7789,17.7322
"Christian ""Shan"" Hendricks Vegetable Market",Saint Croix,Christiansted,Virgin Islands,nan,-64.7043,17.7449
El Maercado Familiar,Arroyo zona urbana,Arroyo,Puerto Rico,00714,-66.0617,17.9686
El Mercado Familiar,Santa Isabel,Santa Isabel,Puerto Rico,00757,-66.4184,17.9723
El Mercado Familiar,Salinas zona urbana,Salinas,Puerto Rico,00751,-66.2954,17.9782
El Mercado Familiar,Guayama,Guayama,Puerto Rico,nan,-66.0938,17.9792
El Mercado Familiar,Guanica,Guanica,Puerto Rico,00653,-66.9205,17.9854
El Mercado Familiar,Patillas,Patillas,Puerto Rico,00723,-66.0135,18.0069
El Mercado Familiar,Peñuelas,Penuelas,Puerto Rico,00624,-66.3572,18.0089


Oops, that didn't answer our question because we sorted from smallest to largest latitude. To look at the largest latitudes, we'll have to sort in reverse order.

In [143]:
farmers_markets_locations.sort('y', descending=True)

MarketName,city,County,State,zip,x,y
Tanana Valley Farmers Market,Fairbanks,Fairbanks North Star,Alaska,99709,-147.781,64.8628
Ester Community Market,Ester,Fairbanks North Star,Alaska,99725,-148.01,64.8459
Fairbanks Downtown Market,Fairbanks,Fairbanks North Star,Alaska,99701,-147.72,64.8444
Nenana Open Air Market,Nenana,nan,Alaska,99704,-149.096,64.5566
Highway's End Farmers' Market,Delta Junction,Fairbanks North Star,Alaska,99737,-145.733,64.0385
MountainTraders,Talkeetna,Matanuska-Susitna,Alaska,99676,-150.118,62.3231
Talkeetna Farmers Market,Talkeetna,Matanuska-Susitna,Alaska,99676,-150.118,62.3228
Denali Farmers Market,Anchorage,nan,Alaska,nan,-150.234,62.3163
Kenny Lake Harvest II,Valdez,nan,Alaska,99686,-145.476,62.1079
Copper Valley Community Market,Copper Valley,nan,Alaska,99737,-145.444,62.0879


(The `descending=True` bit is called an *optional argument*. It has a default value of `False`, so when you explicitly tell the function `descending=True`, then the function will sort in descending order.)

### `sort`

Some details about sort:

1. The first argument to `sort` is the name of a column to sort by.
2. If the column has text in it, `sort` will sort alphabetically; if the column has numbers, it will sort numerically.
3. The value of `farmers_markets_locations.sort("y")` is a *copy* of `farmers_markets_locations`; the `farmers_markets_locations` table doesn't get modified. For example, if we called `farmers_markets_locations.sort("y")`, then running `farmers_markets_locations` by itself would still return the unsorted table.
4. Rows always stick together when a table is sorted.  It wouldn't make sense to sort just one column and leave the other columns alone.  For example, in this case, if we sorted just the `y` column, the farmers' markets would all end up with the wrong latitudes.

**Question 3.3.** Create a version of `farmers_markets_locations` that's sorted by **longitude (`x`)**, with the largest longitudes first.  Call it `farmers_markets_locations_by_longitude`.

In [144]:
farmers_markets_locations_by_longitude = farmers_markets.sort("x", descending=True)
farmers_markets_locations_by_longitude

FMID,MarketName,street,city,County,State,zip,x,y,Website,Facebook,Twitter,Youtube,OtherMedia,Organic,Tofu,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,WildHarvested,updateTime,Location,Credit,WIC,WICcash,SFMNP,SNAP,Season1Date,Season1Time,Season2Date,Season2Time,Season3Date,Season3Time,Season4Date,Season4Time
1004972,"Christian ""Shan"" Hendricks Vegetable Market",Company Street,Saint Croix,Christiansted,Virgin Islands,nan,-64.7043,17.7449,nan,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2009,nan,N,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
1004880,La Reine Farmers Market,Estate Body Slob,Saint Croix,Christiansted,Virgin Islands,nan,-64.7789,17.7322,nan,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2009,nan,N,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
1004881,Anne Heyliger Vegetable Market,Queen Cross Street,Saint Croix,Frederiksted,Virgin Islands,nan,-64.8799,17.7099,nan,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2009,nan,N,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
1004967,Rothschild Francis Vegetable Market,Charlotte Amalie,St. Thomas,St. Thomas,Virgin Islands,nan,-64.9326,18.3428,nan,nan,nan,nan,nan,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2009,nan,N,N,N,N,N,nan,nan,nan,nan,nan,nan,nan,nan
1011914,Feria Agrícola de Luquillo,Centro de Arte y Cultura de Luquillo,Luquillo,Luquillo,Puerto Rico,00773,-65.7207,18.3782,http://www.luquillofarmersmarket.org,www.facebook.com/groups/feriaagricoladeluquillo/,nan,nan,nan,-,N,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,N,N,N,N,Y,N,Y,Y,Y,N,Y,Y,Y,N,Y,N,N,N,5/2/2016 8:18:31 AM,nan,Y,N,N,N,N,01/01/2016 to 12/31/2016,Sat: 8:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011625,El Mercado Familiar,nan,San Lorenzo,San Lorenzo,Puerto Rico,00754,-65.9674,18.1871,http://agricultura.pr.gov,nan,nan,nan,nan,N,N,N,Y,N,N,N,N,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,Y,Y,Y,N,N,N,N,N,8/4/2015 11:46:17 AM,nan,Y,N,Y,Y,N,01/01/2015 to 12/31/2015,Fri: 7:00 AM-1:00 PM;Sat: 7:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011623,El Mercado Familiar,nan,Gurabo,Gurabo,Puerto Rico,00778,-65.9786,18.2526,http://agricultura.pr.gov,nan,nan,nan,nan,N,N,N,Y,N,N,N,N,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,Y,N,Y,N,N,N,N,N,8/4/2015 11:42:03 AM,nan,Y,N,Y,Y,N,01/01/2015 to 12/31/2015,Fri: 7:00 AM-1:00 PM;Sat: 7:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011617,El Mercado Familiar,Estacionamiento del Parque Angelita Lind,Patillas,Patillas,Puerto Rico,00723,-66.0135,18.0069,http://agricultura.pr.gov,nan,nan,nan,nan,N,N,N,Y,N,N,N,N,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,8/4/2015 11:13:20 AM,nan,Y,N,Y,Y,N,01/01/2015 to 12/31/2015,Fri: 7:00 AM-1:00 PM;Sat: 7:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011621,El Mercado Familiar,nan,Caguas zona urbana,Caguas,Puerto Rico,00725,-66.039,18.2324,nan,nan,nan,nan,nan,N,N,N,Y,N,N,N,N,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,Y,N,Y,N,N,N,N,N,8/4/2015 11:30:04 AM,nan,Y,N,Y,Y,N,01/01/2015 to 12/31/2015,Fri: 7:00 AM-1:00 PM;Sat: 7:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan
1011612,El Maercado Familiar,nan,Arroyo zona urbana,Arroyo,Puerto Rico,00714,-66.0617,17.9686,nan,nan,nan,nan,nan,N,N,N,Y,N,N,N,N,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,8/4/2015 10:54:40 AM,nan,Y,N,Y,Y,N,01/01/2015 to 12/31/2015,Fri: 7:00 AM-1:00 PM;Sat: 7:00 AM-1:00 PM;,nan,nan,nan,nan,nan,nan


Now let's say we want a table of all farmers' markets in North Carolina. Sorting won't help us much here because North Carolina is closer to the middle of the dataset.

Instead, we use the table method `where`.

In [145]:
nc_farmers_markets = farmers_markets_locations.where('State', are.equal_to('North Carolina'))
nc_farmers_markets

MarketName,city,County,State,zip,x,y
Afton Village Farmers Market,Concord,Cabarrus,North Carolina,28027,-80.6702,35.414
Alamance County Farmers Market,Burlington,Alamance,North Carolina,27215,-79.4357,36.0943
Alexander County Farmers Market,Taylorsville,Alexander,North Carolina,28681,-81.1781,35.9197
Alleghany Farmers Market,Sparta,Alleghany,North Carolina,28675,-81.1226,36.503
Andrews Farmers Market,Andrews,Cherokee,North Carolina,nan,-83.8232,35.2027
Anson County Farmers Market,Wadesboro,Anson,North Carolina,28170,-80.0526,34.9408
Apex Farmers Market,Apex,Wake,North Carolina,27502,-78.8499,35.732
Ashboro Downtown Farmers Market,Asheboro,Randolph,North Carolina,28801,-79.8175,35.7049
Ashe County Farmers Market,West Jefferson,Ashe,North Carolina,28694,-81.4935,36.4025
Asheville City Market,Asheville,Buncombe,North Carolina,28801,-82.5489,35.5935


Ignore the syntax for the moment.  Instead, try to read that line like this:

> Assign the name **`nc_farmers_markets`** to a table whose rows are the rows in the **`farmers_markets_locations`** table **`where`** the **`'State'`**s **`are` `equal` `to` `North Carolina`**.

### `where`

Now let's dive into the details a bit more.  `where` takes 2 arguments:

1. The name of a column.  `where` finds rows where that column's values meet some criterion.
2. A predicate that describes the criterion that the column needs to meet.

The predicate in the example above called the function `are.equal_to` with the value we wanted, 'North Carolina'.  We'll see other predicates soon. 

`where` returns a table that's a copy of the original table, but **with only the rows that meet the given predicate**.

**Question 3.4.** Use `nc_farmers_markets` to create a table called `oc_markets` containing farmers' markets in `Orange` County, North Carolina. 

*Note:* You may see some inconsistency here in the data. Most farmer's markets in Orange County have the `County` value of `Orange`, while one farmers market in Orange County has the `County` value of `Orange County`. You can only consider those with the value `Orange`, or you can come up with a way to include both.

In [146]:
oc_markets = nc_farmers_markets.where("County", are.contained_in(["Orange", "Orange County"]))
oc_markets

MarketName,city,County,State,zip,x,y
Carrboro Farmers' Market,Carrboro,Orange,North Carolina,27510,-79.0776,35.9109
EFLAND FARMERS MARKET,Efland,Orange County,North Carolina,nan,-79.169,36.0812
Eno River Farmers Market in Downtown Hillsborough,Hillsborough,Orange,North Carolina,27278,-79.0978,36.074
Hillsborough Farmers Market,Hillsborough,Orange,North Carolina,27278,-79.0807,36.0555
Southern Village Farmers Market,Chapel Hill,Orange,North Carolina,27517,-79.0659,35.881
The Chapel Hill Farmers' Market,Chapel Hill,Orange,North Carolina,27514,-79.0294,35.9275


Recognize any of them?

# 4. Importing code

![imports](https://external-preview.redd.it/ZVPjiFo_Ubl4JeiU63SaTjdIoq5zveSnNZimKpgn2I8.png?auto=webp&s=bf32c94b630befa121075c1ae99b2599af6dedc5) 

[source](https://www.reddit.com/r/ProgrammerHumor/comments/cgtk7s/theres_no_need_to_reinvent_the_wheel_oc/)

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to **import modules**. A module is a file with Python code that has defined variables and functions. By importing a module, we are able to use its code in our own notebook.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with a radius of 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [147]:
# Just run this cell

import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

78.53981633974483

In the code above, the line `import math` imports the math module. This statement creates a module and then assigns the name `math` to that module. We are now able to access any variables or functions defined within `math` by typing the name of the module followed by a dot, then followed by the name of the variable or function we want.

    <module name>.<name>

**Question 4.1.** The module `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71. Compute $\sqrt{e^{2\pi}+30\pi}$, giving it the name `near_twentyfive`.

*Remember: You can access `pi` from the `math` module as well!*

In [148]:
near_twentyfive = math.sqrt(math.pow(math.e, 2*math.pi) + 30*math.pi)
near_twentyfive

25.094609682807544

## 4.1. Accessing functions

In the question above, you accessed variables within the `math` module. 

**Modules** also define **functions**.  For example, `math` provides the name `log` for the natural log function. Having imported `math` already, we can write `math.log(3)` to compute the natural log of 3. 

For your reference, below are some more examples of functions from the `math` module.

Notice how different functions take in different numbers of arguments. Often, the [documentation](https://docs.python.org/3/library/math.html) of the module will provide information on how many arguments are required for each function.

In [149]:
# Calculating logarithms (the logarithm of 8 in base 2).
# The result is 3 because 2 to the power of 3 is 8.
math.log(8, 2)

3.0

In [150]:
# Calculating square roots.
math.sqrt(5)

2.23606797749979

There are various ways to import and access code from outside sources. The method we used above — `import <module_name>` — imports the entire module and requires that we use `<module_name>.<name>` to access its code. 

We can also import a specific constant or function instead of the entire module. Notice that you don't have to use the module name beforehand to reference that particular value. However, you do have to be careful about reassigning the names of the constants or functions to other values!

In [151]:
# Importing just cos and pi from math.
# We don't have to use `math.` in front of cos or pi
from math import cos, pi
print(cos(pi))

# We do have to use it in front of other functions from math, though
math.log(pi)

-1.0


1.1447298858494002

Or we can import every function and value from the entire module.

In [152]:
# Lastly, we can import everything from math using the *
# Once again, we don't have to use 'math.' beforehand 
from math import *
log(pi)

1.1447298858494002

Don't worry too much about which type of import to use. It's often a coding style choice left up to each programmer. In this course, you'll always import the necessary modules when you run the setup cell (like the first code cell in this lab).

Let's move on to practicing some of the table operations you've learned in lecture!

# 5. More Table operations

So far in lab we've only been using `where` with the predicate that requires finding the values in a column to be *exactly* equal to a certain value. However, there are many other predicates. Here are a few:

|Predicate|Example|Result|
|-|-|-|
|`are.equal_to`|`are.equal_to(50)`|Find rows with values equal to 50|
|`are.not_equal_to`|`are.not_equal_to(50)`|Find rows with values not equal to 50|
|`are.above`|`are.above(50)`|Find rows with values above (and not equal to) 50|
|`are.above_or_equal_to`|`are.above_or_equal_to(50)`|Find rows with values above 50 or equal to 50|
|`are.below`|`are.below(50)`|Find rows with values below 50|
|`are.between`|`are.between(2, 10)`|Find rows with values above or equal to 2 and below 10|

Now that you're familiar with table operations, let’s answer more questions using this data!

Often, we want to perform multiple operations - sorting, filtering, or others - in order to turn a table we have into something more useful. You can do these operations one by one, e.g.

```
first_step = original_tbl.where(“col1”, are.equal_to(12))
second_step = first_step.sort(‘col2’, descending=True)
```

However, since the value of the expression `original_tbl.where(“col1”, are.equal_to(12))` is itself a table, you can just call a table method on it:

```
original_tbl.where(“col1”, are.equal_to(12)).sort(‘col2’, descending=True)
```
You should organize your work in the way that makes the most sense to you, using informative names for any intermediate tables you create. 

**Question 5.1.** Suppose that you would like to visit a farmers' market in Orange County, NC that sells wine, soap, coffee, and takes credit cards (`Credit`). Use the `farmers_markets` table to produce a new table `shopping_trip` that can help you to succinctly decide where to go.

In [153]:
shopping_trip = farmers_markets.where("State", "North Carolina").where("County", are.contained_in(["Orange", "Orange County"])).where("Wine", "Y").where("Soap", "Y").where("Coffee", "Y").where("Credit", "Y")
shopping_trip

FMID,MarketName,street,city,County,State,zip,x,y,Website,Facebook,Twitter,Youtube,OtherMedia,Organic,Tofu,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,WildHarvested,updateTime,Location,Credit,WIC,WICcash,SFMNP,SNAP,Season1Date,Season1Time,Season2Date,Season2Time,Season3Date,Season3Time,Season4Date,Season4Time
1011740,Carrboro Farmers' Market,301 West Main Street,Carrboro,Orange,North Carolina,27510,-79.0776,35.9109,http://www.carrborofarmersmarket.com,https://www.facebook.com/carrborofarmersmarket,https://twitter.com/carrboromarket,nan,nan,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,12/10/2015 1:27:13 PM,nan,Y,Y,N,Y,Y,11/07/2015 to 03/26/2016,Sat: 9:00 AM-12:00 PM;,04/02/2016 to 10/29/2016,Sat: 7:00 AM-12:00 PM;,04/06/2016 to 11/16/2016,Wed: 3:00 PM-6:00 PM;,nan,nan
1000175,Eno River Farmers Market in Downtown Hillsborough,144 E. Margaret Lane,Hillsborough,Orange,North Carolina,27278,-79.0978,36.074,http://www.enoriverfarmersmarket.com,https://www.facebook.com/pages/Eno-River-Farmers-Market- ...,https://twitter.com/EnoRiverFM,nan,nan,Y,N,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,N,Y,N,N,Y,Y,Y,Y,Y,Y,Y,N,Y,N,N,Y,N,Y,6/28/2016 11:36:58 AM,nan,Y,Y,Y,N,N,04/01/2016 to,Sat: 8:00 AM-12:00 PM;,11/05/2016 to 03/25/2017,Sat: 10:00 AM-12:00 PM;,nan,nan,nan,nan
1004439,The Chapel Hill Farmers' Market,201 S. Estes Dr,Chapel Hill,Orange,North Carolina,27514,-79.0294,35.9275,http://www.thechapelhillfarmersmarket.com,https://www.facebook.com/TheChapelHillFarmersMarket,https://twitter.com/chfarmersmarket,nan,pinterest: chfarmersmarket instagram: chfarmersmarket,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,N,N,Y,N,Y,3/31/2014 4:42:09 PM,Private business parking lot,Y,N,N,N,N,03/29/2014 to 11/27/2014,Sat: 8:00 AM-12:00 PM;,04/15/2013 to 11/26/2013,Tue: 3:00 PM-6:00 PM;,11/28/2014 to 03/21/2015,Sat: 9:00 AM-12:00 PM;,nan,nan


**Question 5.2.** Which farmers market (or markets) could you choose for this shopping trip based on the table that you constructed in the previous question?

*Carrboro Farmers' Market* or *Eno River Farmers Market in Downtown Hillsborough* or *The Chapel Hill Farmers' Market*

**Question 5.3.** Use `num_rows` (and arithmetic) to find the *proportion* of farmers' markets in North Carolina that sell wine (`Wine`). Assign `NC_wine_prop` to this proportion

How does this compare the *proportion of* of all farmers' markets in the `farmers_markets` table that sell wine? Assign `US_wine_prop` to this proportion

In [154]:
NC_wine_prop = farmers_markets.where("State", "North Carolina").where("Wine", "Y").num_rows / farmers_markets.where("State", "North Carolina").num_rows
US_wine_prop = farmers_markets.where("Wine", "Y").num_rows / farmers_markets.num_rows

print("Proportion of North Carolina farmers' markets selling wine:", NC_wine_prop)
print("Proportion of US farmers' markets selling wine:", US_wine_prop)

Proportion of North Carolina farmers' markets selling wine: 0.12301587301587301
Proportion of US farmers' markets selling wine: 0.10741867540369764


**Question 5.4.** Find a state where the proportion of farmers' markets selling wine is more than double the national average (`US_wine_prop`) and determine this proportion for your chosen state.

In [155]:
import numpy as np

Chosen_State_Array = {}

# national proportion
us_wine_prop = farmers_markets.where("Wine", "Y").num_rows / farmers_markets.num_rows

# Get unique states
unique_states = np.unique(farmers_markets.column("State"))

for state in unique_states:
    state_markets = farmers_markets.where("State", state)
    state_wine_prop = state_markets.where("Wine", "Y").num_rows / state_markets.num_rows
    
    if state_wine_prop > 2 * us_wine_prop:
        Chosen_State_Array[state] = state_wine_prop



Chosen_State_Array
Chosen_State = max(Chosen_State_Array, key=Chosen_State_Array.get)
Chosen_State

'Vermont'

In [156]:
Chosen_State_wine_prop = Chosen_State_Array[Chosen_State]
print("Proportion of", Chosen_State, "farmers' markets selling wine:", Chosen_State_wine_prop)

Proportion of Vermont farmers' markets selling wine: 0.3229166666666667


# 6. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day (That's if you're pretty fast at doing arithmetic in your head!).

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying each number in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**. 

## 6.1. Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [157]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

**Question 6.1.1.** Make an array containing the numbers $\pi$, 1, -1, 0, and $e$, in that order.  Name it `interesting_numbers`.  

*Hint:* How have you found the values $\pi$ and $e$ earlier in this lab?

In [158]:
interesting_numbers = make_array(math.pi, 1, -1, 0, math.e)
interesting_numbers

array([ 3.14159265,  1.        , -1.        ,  0.        ,  2.71828183])

**Question 6.1.2.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.

In [159]:
hello_world_components = make_array("Hello", ",", " ", "world", "1")
hello_world_components

array(['Hello', ',', ' ', 'world', '1'],
      dtype='<U5')

###  `np.arange`
Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Run the following cells to see some examples!

In [160]:
import numpy as np

# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [161]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

**Question 6.1.3.** Create an array with the multiples of 34 from 0 up to (**and including**) 3434. Its elements should be 0, 34, 68, 102, etc.

In [162]:
multiples_of_34 = np.arange(0, 3435, 34)
multiples_of_34

array([   0,   34,   68,  102,  136,  170,  204,  238,  272,  306,  340,
        374,  408,  442,  476,  510,  544,  578,  612,  646,  680,  714,
        748,  782,  816,  850,  884,  918,  952,  986, 1020, 1054, 1088,
       1122, 1156, 1190, 1224, 1258, 1292, 1326, 1360, 1394, 1428, 1462,
       1496, 1530, 1564, 1598, 1632, 1666, 1700, 1734, 1768, 1802, 1836,
       1870, 1904, 1938, 1972, 2006, 2040, 2074, 2108, 2142, 2176, 2210,
       2244, 2278, 2312, 2346, 2380, 2414, 2448, 2482, 2516, 2550, 2584,
       2618, 2652, 2686, 2720, 2754, 2788, 2822, 2856, 2890, 2924, 2958,
       2992, 3026, 3060, 3094, 3128, 3162, 3196, 3230, 3264, 3298, 3332,
       3366, 3400, 3434])

##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Raleigh-Durham International Airport site for the Fall 2024 semester, from the first day of class on August 19, 2024 through commencement on December 15, 2024.  To analyze the data, we want to know when each reading was taken, but we find that the data do not include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of August 19th 2024 (midnight on August 19th) and each subsequent reading was taken exactly 1 hour after the last.

**Question 6.1.4.** Create an array of the time, *in seconds*, since the start of August 19th 2024, and ending before the start of December 16th 2024 at which each hourly reading was taken.  Name it `collection_times`.

*Hint 1:* How many days are in each month? There are 30 days in September, which is equivalent to ($30 \times 24$) hours or ($30 \times 24 \times 60 \times 60$) seconds.

In [163]:
total_seconds = 119 * 24 * 60 * 60
hour_in_seconds = 60 * 60 
collection_times = np.arange(0, total_seconds, hour_in_seconds)
collection_times

array([       0,     3600,     7200, ..., 10270800, 10274400, 10278000])

## 6.2. Working with single elements of arrays ("indexing")
The next cell creates an array called `NCpop` that includes estimated North Carolina populations in every year from 1900 to roughly the 2021. The estimates come from the [FRED Economic Data](https://fred.stlouisfed.org/series/NCPOP).


In [164]:
NCpop = Table.read_table("../../shared/Data/NCpop.csv").column("NCPOP")
NCpop

array([ 1897000,  1926000,  1956000,  1986000,  2017000,  2051000,
        2077000,  2105000,  2142000,  2174000,  2221000,  2276000,
        2313000,  2362000,  2421000,  2473000,  2513000,  2546000,
        2522000,  2535000,  2588000,  2651000,  2700000,  2761000,
        2830000,  2895000,  2959000,  3027000,  3082000,  3133000,
        3167000,  3184000,  3227000,  3268000,  3304000,  3323000,
        3346000,  3385000,  3440000,  3514000,  3574000,  3589000,
        3569000,  3654000,  3560000,  3533000,  3706000,  3769000,
        3837000,  3911000,  4068000,  4120000,  4109000,  4120000,
        4131000,  4242000,  4309000,  4368000,  4376000,  4458000,
        4573000,  4663000,  4707000,  4742000,  4802000,  4863000,
        4896000,  4952000,  5004000,  5031000,  5084411,  5203531,
        5301150,  5389852,  5470911,  5547188,  5607964,  5685607,
        5759492,  5823491,  5898980,  5956653,  6019101,  6077056,
        6164006,  6253954,  6321578,  6403700,  6480594,  6565

Here's how we get the first element of `NCpop`, which is the NC population in the first year in the dataset, 1900.

In [165]:
NCpop.item(0)

1897000

The value of that expression is the number 1897000 (almost 2 million), because that's the first thing in the array `NCpop`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is a weird convention in computer science.  0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `NCpop`.  Read and run each cell.

In [166]:
# The 13th element in the array is the population
# in 1912 (which is 1900 + 12).
nc_pop_1912 = NCpop.item(12)
nc_pop_1912

2313000

In [167]:
# The 66th element is the population in 1965.
nc_pop_1965 = NCpop.item(65)
nc_pop_1965

4863000

In [168]:
# The array has only 122 elements, so this doesn't work.
# (There's no element with 122 other elements before it.)
nc_pop_2022 = NCpop.item(122)
nc_pop_2022

IndexError: index 122 is out of bounds for axis 0 with size 122

Since `make_array` returns an array, we can call `.item(3)` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [169]:
make_array(-1, -3, 4, -2).item(3)

-2

**Question 6.2.1.** Set `population_1979` to the North Carolina population in 1979, by getting the appropriate element from `NCpop` using `item`.

In [170]:
population_1979 = NCpop.item(79)
population_1979

5823491

## 6.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Logarithms
Here is one simple question we might ask about NC population:

> How big was the population in *orders of magnitude* in each year?

Orders of magnitude quantify how big a number is by representing it as the power of another number (for example, representing 104 as $10^{2.017033}$). One way to do this is by using the logarithm function. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [171]:
nc_pop_1900_magnitude = math.log10(NCpop.item(0))
nc_pop_1901_magnitude = math.log10(NCpop.item(1))
nc_pop_1902_magnitude = math.log10(NCpop.item(2))
nc_pop_1903_magnitude = math.log10(NCpop.item(3))
...

Ellipsis

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

**Question 6.3.1.** Use `np.log10` to compute the logarithms of the North Carolina population in every year.  Give the result (an array of 122 numbers) the name `NCpop_magnitudes`.  Your code should be very short.

In [172]:
NCpop_magnitudes = np.log10(NCpop)
NCpop_magnitudes

array([ 6.27806733,  6.28465628,  6.29136885,  6.29797924,  6.3047059 ,
        6.31196566,  6.3174365 ,  6.3232521 ,  6.33081947,  6.33725954,
        6.34654856,  6.35717226,  6.36417563,  6.37327989,  6.38399479,
        6.39322412,  6.40019249,  6.4058584 ,  6.40174508,  6.40397796,
        6.41296427,  6.42340973,  6.43136376,  6.44106641,  6.45178644,
        6.46164857,  6.47114497,  6.48101242,  6.48883263,  6.49596039,
        6.50064806,  6.50297306,  6.50879897,  6.51428205,  6.51904004,
        6.52153034,  6.52452594,  6.52955867,  6.53655844,  6.54580176,
        6.55315455,  6.55497346,  6.55254655,  6.56276854,  6.55145   ,
        6.54814364,  6.56890541,  6.57622614,  6.5839918 ,  6.59228782,
        6.60938094,  6.61489722,  6.61373614,  6.61489722,  6.61605519,
        6.62757066,  6.63437649,  6.64028263,  6.64107731,  6.64914006,
        6.6602012 ,  6.66866542,  6.6727442 ,  6.67596155,  6.68142216,
        6.68690427,  6.68984141,  6.69478064,  6.6993173 ,  6.70

What you just did is called *elementwise* application of `np.log10`, since `np.log10` operates separately on each element of the array that it's called on. The textbook's [section](https://www.inferentialthinking.com/chapters/05/1/Arrays)  on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

##### Arithmetic
Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 million to get numbers in millions:

In [173]:
NCpop_in_millions = NCpop / 1000000
NCpop_in_millions

array([  1.897   ,   1.926   ,   1.956   ,   1.986   ,   2.017   ,
         2.051   ,   2.077   ,   2.105   ,   2.142   ,   2.174   ,
         2.221   ,   2.276   ,   2.313   ,   2.362   ,   2.421   ,
         2.473   ,   2.513   ,   2.546   ,   2.522   ,   2.535   ,
         2.588   ,   2.651   ,   2.7     ,   2.761   ,   2.83    ,
         2.895   ,   2.959   ,   3.027   ,   3.082   ,   3.133   ,
         3.167   ,   3.184   ,   3.227   ,   3.268   ,   3.304   ,
         3.323   ,   3.346   ,   3.385   ,   3.44    ,   3.514   ,
         3.574   ,   3.589   ,   3.569   ,   3.654   ,   3.56    ,
         3.533   ,   3.706   ,   3.769   ,   3.837   ,   3.911   ,
         4.068   ,   4.12    ,   4.109   ,   4.12    ,   4.131   ,
         4.242   ,   4.309   ,   4.368   ,   4.376   ,   4.458   ,
         4.573   ,   4.663   ,   4.707   ,   4.742   ,   4.802   ,
         4.863   ,   4.896   ,   4.952   ,   5.004   ,   5.031   ,
         5.084411,   5.203531,   5.30115 ,   5.389852,   5.470

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [174]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [ 20.12  39.9   31.01]
Tips:			 [ 4.024  7.98   6.202]


**Question 6.3.2.** Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.

In [175]:
total_charges = 1.2 * restaurant_bills
total_charges

array([ 24.144,  47.88 ,  37.212])

**Question 6.3.3.** The array `more_restaurant_bills_array` contains 100,000 bills!  Compute the total charge for each one.  How is your code different?

In [176]:
more_restaurant_bills_array = Table.read_table("../../shared/Data/more_restaurant_bills.csv").column("Bill")
more_total_charges = 1.2 * more_restaurant_bills_array
more_total_charges

array([ 20.244,  20.892,  12.216, ...,  19.308,  18.336,  35.664])

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 6.3.4.** What was the sum of all the total bills in `more_restaurant_bills_array`, *including tips*? Round your answer to the nearest cent. We will discuss in class why you see extra decimal places even though you rounded to fewer.

In [177]:
sum_of_bills = round(sum(more_total_charges), 2)
sum_of_bills

1795730.0600000001